File: library/swift/src/StreamCallback.swift
import Foundation

/// Callback interface for receiving stream events.
@objc
public protocol StreamCallback {
  /// Invoked when response headers are read.
  ///
  /// - parameter headers:    The headers of the response.
  /// - parameter statusCode: The status code of the response.
  func onHeaders(_ headers: [String: [String]], statusCode: Int)

  /// Invoked when response body is read.
  ///
  /// - parameter data:      Bytes in the response.
  /// - parameter endStream: True if the stream is complete.
  func onData(_ data: Data, endStream: Bool)

  /// Invoked when response metadata is read.
  ///
  /// - parameter metadata:  The metadata of the response.
  /// - parameter endStream: True if the stream is complete.
  func onMetadata(_ metadata: [String: [String]], endStream: Bool)

  /// Invoked when response trailers are read.
  ///
  /// - parameter trailers: The trailers of the response.
  func onTrailers(_ trailers: [String: [String]])

  /// Invoked when there is an internal Envoy error with the stream.
  ///
  /// - parameter error: The error that occurred with the stream.
  func onError(_ error: Error)

  /// Invoked when the stream is canceled.
  func onCanceled()

  /// Invoked when the stream is completed.
  func onCompletion()
}

File: library/kotlin/src/io/envoyproxy/envoymobile/StreamCallback.kt
package io.envoyproxy.envoymobile

import java.nio.ByteBuffer

interface StreamCallback {
  /**
   * Called when a response headers is received by the stream.
   *
   * @param headers the headers of the response.
   * @param statusCode the status code of the response.
   */
  fun onHeaders(headers: Map<String, List<String>>, statusCode: Int)

  /**
   * Called when a data frame is received by the stream.
   *
   * @param byteBuffer the byte buffer of the response.
   * @param endStream true if the stream is complete.
   */
  fun onData(byteBuffer: ByteBuffer, endStream: Boolean)

  /**
   * Called when a response metadata has been received by the stream.
   *
   * @param metadata the metadata of a response.
   * @param endStream true if the stream is complete.
   */
  fun onMetadata(metadata: Map<String, List<String>>, endStream: Boolean)

  /**
   * Invoked whenever a response trailers have been read.
   *
   * @param trailers the trailers of the response.
   */
  fun onTrailers(trailers: Map<String, List<String>>)

  /**
   * Called when an internal Envoy exception occurs with the stream.
   *
   * @param envoyException the exception associated with the stream.
   */
  fun onError(envoyException: EnvoyException)

  /**
   * Called when the stream is cancelled.
   *
   */
  fun onCanceled()

  /**
   * Called when the stream has been completed.
   *
   */
  fun onCompletion()
}
